core/src/com/unciv/logic/city/CityInfo.kt

```diff
@@ -175,60 +180,85 @@ class CityInfo {
         cityConstructions.addFreeBuildings()
     }
 
-    private fun setNewCityName(civInfo: CivilizationInfo) {
-        val nationCities = civInfo.nation.cities
-        val cityNameIndex = civInfo.citiesCreated % nationCities.size
-        val cityName = nationCities[cityNameIndex]
-
-        val cityNameRounds = civInfo.citiesCreated / nationCities.size
-        if (cityNameRounds > 0 && civInfo.hasUnique(UniqueType.BorrowsCityNames)) {
-            name = borrowCityName()
-            return
-        }
-        val cityNamePrefix = when (cityNameRounds) {
-            0 -> ""
-            1 -> "New "
-            else -> "Neo "
+    /**
+     * Generates and returns a new city name.
+     *
+     * This method attempts to return the first unused name in the given [nationCityNames] taking
+     * [usedCityNames] into consideration. If that fails, it then checks whether the nation
+     * [hasBorrowsCityNames] and, if true, returns a borrowed name. Else, it repeatedly attaches a
+     * predefined prefix to the list of names until an unused name is successfully generated.
+     *
+     * @param nationCityNames Every city name of the nation.
+     * @param usedCityNames Every city name in use including foreign cities.
+     * @param hasBorrowsCityNames Whether this nation borrows city names or not
+     * ([UniqueType.BorrowsCityNames]).
+     * @return A new city name in [String].
+     */
+    private fun generateNewCityName(
+        nationCityNames: List<String>, usedCityNames: List<String>, hasBorrowsCityNames: Boolean
+    ): String {
+        // Attempt to return the first missing name from the list of city names
+        // ex) Rome, Antium, ... , Interrama, Adria
+        nationCityNames.forEach { cityName -> if (cityName !in usedCityNames) { return cityName } }
+
+        // If all names are taken and this nation has the BorrowsCityNames unique,
+        // return a random borrowed city name
+        if (hasBorrowsCityNames) return generateBorrowedCityName(usedCityNames)
+
+        /*
+        If the nation doesn't have the unique above,
+        return the first missing name with an increasing number of prefixes attached
+        ex) New Rome, New Antium, ... , New Adria, New New Rome, New New Antium, ...
+        TODO: Make the prefix moddable per nation?
+         Also, if then, what about other languages using suffixes?
+        TODO: Could've used a [sequence] from 1 to [int.MAX_VALUE] instead of the while loop but my
+         IDE wouldn't stop complaining about a return expression being required. I wonder whether
+         this is the best or there is a better way to do this.
+        */
+        val prefix = "New "
+        var candidate: String?
+        var number = 1
+        while (true) {
```

Comment:
I'm a bit hesitant with `while (true)` loops as for some reason they tend to repeat indefinitely even when that should be impossible. Maybe a for loop repeating 10 times and otherwise returning `The City without a Name` as used in `generateBorrowedCityName`? Having a city called "New New New New New New New New New New New New Rome" probably also isn't exactly great either imo

Reply:
> The old method allowed infinite duplicates of "Neo (name here)". The reworked method would instead generate "New New (name here)", "New New New (name here)" and so on, hence ensuring that every city name is unique.
> \- https://github.com/yairm210/Unciv/pull/6439#issuecomment-1079749064

On a second thought, I think I'll just use a list of prefixes as you and @SomeTroglodyte advised...